### ISSUE TITLE
Follow up on Shuffler interacted with DataLoader

### ISSUE THREAD

**Issue by ejguan** [pytorch/data#426]
### 🚀 The feature

After https://github.com/pytorch/pytorch/pull/77741 and https://github.com/pytorch/pytorch/pull/77855 are landed, there are two follow-up discussions on the shuffling behavior of `DataLoader` and `IterDataPipe`.
- For Distributed training, we currently have to let users to call `torch.manual_seed()` with the same seed across distributed processes to guarantee all shuffles have the same effect for `DataPipe`. However, for `Dataset`, users can provide the seed to`DistributedSampler` to control shuffle order. Side note: the seeding strategy for `DistributedSampler` is also bad as `seed + epoch`. For `IterableDataset`, we don't have the functionality to shuffle it in either distirbuted or mp environment.
  - Potential solution for `DataPipe`: We could check the environment whether or not in distributed setting (`torch.distributed. is_initialized()`. And, We could transmit seeds from the `DataLoader` in Rank 0 to other `DataLloader`s. And each `DataLoader` in distributed process needs to wait until this environment variable is set, then it should be able to set the same seed for all `Shuffler`s.
- For persistent workers, , each persistent worker process would set the same seed to `Shuffler`s deterministically. However, users won't be able to reset seeds for persistent workers after the first iteration starts via `torch.manual_seed` (The `seed` and RNG are persistent in worker process). But, for `Dataset`, users still can reset seed in the main process via `Sampler`.
  - Potential solution: Use [`_ResumeIteration`](https://github.com/pytorch/pytorch/blob/97fa1d317f68fec5c99eb4144d01795d3fb2a9e1/torch/utils/data/_utils/worker.py#L117-L120) to pass seed from main process to persistent process.

### Motivation, pitch

- Make `Shuffler` automatically seeded in distributed training
- Make `Shuffler` can be seeded per epoch even with persistent workers

### Additional context

cc: @NicolasHug @VitalyFedyunin

**Comment by ejguan** [comment]
Also cc: @parmeet  about `ShufflerDataPipe` now is deterministic with `DataLoader`. Users could use `torch.manual_seed()` in their training script to control shuffle state.
Even without `torch.manual_seed()`, all `ShufflerDataPipe` in worker processes should have the same random state to ensure the data order is the same before sharding.

**Comment by parmeet** [comment]
This is great! Thanks so much @ejguan for the fix :)

**Comment by ejguan** [comment]
@parmeet 
This issue is used to tracking other two features needed for distributed training.

**Comment by ejguan** [comment]
> I am not sure what these options are and if users really have to specify them explicitly in order to do multiprocessing?

I am using `spawn` simply because it's available across different OSs (easier to write tests lol). Windows doesn't have `fork`.

**Comment by ejguan** [comment]
Closing this Issue as it's resolved by https://github.com/pytorch/pytorch/pull/78765
